### ISSUE TITLE
Site Interactivity breaks sporadically 

### ISSUE THREAD

**Issue by zmanning29** [gridsome/gridsome#1464]
## Description
Website hosted with Netlify, randomly becomes unresponsive and will not load. <g-image> <g-link> components. Navigation updates in the URL bar, but the page never changes or reloads. We are also experiencing the need to double click links before routing occurs. 

### Steps to reproduce
Client is the only one able to reproduce intermittently. Despite my attempts to reproduce using metered web connections, and Browserstack. Any guidance here would be helpful.

**Comment by daanbeverdam** [comment]
I've noticed this behavior in production (deployed to Netlify) when landing on pages that don't exist: 404s. It is not reproducible in a development environment. Could be related to #1168. Hope this helps, let me know.

**Comment by zmanning29** [comment]
I think it's definitely related. I am sure it's probably a combination of the g-image and g-link components. That said. It's so unpredictable that I struggle to reproduce it. When it happens though, it's a catastrophic failure. Nothing works at any level.

**Comment by mklueh** [comment]
Also Check all your HTML code for errors. Some tags should not be child of another and can lead to those issues. If you are using Webstorm those will be highlighted

**Comment by amxmln** [comment]
This sounds like a client-side hydration failure to me, I think there was another issue where this was discussed in the past. Maybe check if you have any `<noscript>` tags containing dynamic statements or block-elements in your code. Removing those fixed similar symptoms for me once.

**Comment by zmanning29** [comment]
I agree @amxmln. I will take a look around. They are using some 3rd party applications that require JavaScript to function so it's likely it could be an issue.

**Comment by mschmidtkorth** [comment]
#650 has a lot of examples as well for failure reasons.

**Comment by rylanharper** [comment]
I recently created (an almost finished) Gridsome Shopify starter.. I had a hydration issue as well where pages would not render upon page refresh or by navigating back to the previous page/link the user had visited.. 

After spending days rebuilding the site from the ground up and making sure it wasn't a third party plugins or my vuex-store I found the issue to be that my `g-links` needed to have an end slash on the route.

Example: `<g-link to="/cart/">Cart</g-link>`

In fact my starter site https://gridsome-headless-shopify.netlify.app/ will still break if you navigate to another page, take off the end slash of the url, and reload lol (lets hope users don't do that in mass though)

Now I know this is in the documentation [here](https://gridsome.org/docs/linking/#linking), but for every other Gridsome site I have made in the past this was not an issue. For whatever reason I needed to have that in this current project.. I also looked all through all the issues in #650 and there is definitely something going, but it is so specific to projects that it is hard to narrow down what is causing it. In general, what I have read in here to be safe is:

- End your `g-links` with an end `/`just in case. I know it looks bad url wise, but f*ck it
- Be sure your component scripts correctly follow the [Vue Lifecycle Hooks ](https://v3.vuejs.org/api/options-lifecycle-hooks.html) (there have been issues with people using simple `mounted` functions that should be `methods`
- DO NOT have a `g-link` within a `g-link`. This will cause the world to literally explode (In terms of Gridsome lol)
- If a component starts with `g-link` as the parent just put a div before it (just in case). I am not sure if this will do anything, but based on issues in #605 it might help?

I hope this helps!

**Comment by mschmidtkorth** [comment]
Good one @rylanharper! Though it seems in my case there is another issue causing this behavior (see [here](https://github.com/mschmidtkorth/gridsome-debug-2)).

**Comment by huntertrammell** [comment]
Has anyone found any workarounds for this when the known culprits aren't present (hydration, v-if, g-link, g-image, netlify-form, etc) or even a way to remove client side routing from gridsome and just allow the server to handle the routing?  Looking for a quick fix to override the page routing while I figure out a long term fix. I really like Gridsome, think it has a lot of potential and would love to contribute but I can't deploy a site with broken routing.

I've done everything I can think of outside just updating each link to href="javascript:void(0)" onclick="window.location='PATH'". Obviously not ideal, but better than no routing at all.

**Comment by cameronrr** [comment]
I haven't come across these issues yet with a couple of sites deployed on Netlify, but maybe some of these notes will help. I also just posted [here](https://github.com/gridsome/gridsome/issues/650#issuecomment-850039671) regarding how to use Netlify forms in case that is the cause of any issues.

Always specify your links per the [docs](https://gridsome.org/docs/linking/#link-between-pages) i.e. with the trailing slash.

I create a [custom 404 page](https://gridsome.org/docs/pages/#custom-404-page). The routes generated by gridsome automatically include a /404/ route as well as a final fallback '*' route to this component.

You shouldn't need to be manually adding Netlify redirects for the standard routes in your project including /404/ and /index.html. This all works out of the box for me. Manual entries would be needed for things like proxies, routing to other hosts, routing from old urls after a site migration/rebuild, etc.

Regarding automated creation of a redirects file for a Netlify deploy if you do have dynamically created/managed pages, the following works for my projects. This creates the _redirects file as part of the final build steps, based on the dynamic pages generated by your project. This can be set in addition to anything you statically specify in a netlify.toml file, which you store in /static prior to the build.

```javascript
//gridsome.server.js

const fs = require('fs')

module.exports = function (api) {

  // ... some dynamic pages using the Pages API i.e. api.createManagedPages() or api.createPages()

  api.afterBuild(({ redirects }) => {
    if (redirects) {
      let rules = []
      for (const rule of redirects) {
        rules.push(`${rule.from}\t${rule.to}\t${rule.status}`)
      }
      fs.appendFileSync('./dist/_redirects', rules.join('\n'))
    }
  }) 
}
